src/Adapter/MSTest.TestAdapter/Helpers/MethodRunner.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Globalization;
+using System.Reflection;
+
+using Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel;
+
+using UnitTestOutcome = Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel.UnitTestOutcome;
+
+namespace Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Helpers;
+
+internal static class MethodRunner
+{
+    internal static TestFailedException? RunWithTimeoutAndCancellation(
+        Action action, CancellationTokenSource cancellationTokenSource, int? timeout, MethodInfo methodInfo,
+        string methodCancelledMessageFormat, string methodTimedOutMessageFormat)
+    {
+        try
+        {
+            var executionTask = Task.Run(action, cancellationTokenSource.Token);
+
+            if (!timeout.HasValue)
+            {
+                executionTask.Wait(cancellationTokenSource.Token);
```

Comment:
also this one can throw OCE but it's not a real timeout

Reply:
OperationCanceledException
The cancellationToken was canceled.

Reply:
Now it's better.